include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -0,0 +1,231 @@
+#ifndef _IRR_BUILTIN_MATERIAL_COMPILER_GLSL_RASTERIZATION_IMPL_INCLUDED_
+#define _IRR_BUILTIN_MATERIAL_COMPILER_GLSL_RASTERIZATION_IMPL_INCLUDED_
+
+#include <irr/builtin/material_compiler/glsl/common.glsl>
+
+void instr_eval_execute(in instr_t instr, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache _uf, inout bool ts_flag)
+{
+	uint op = instr_getOpcode(instr);
+
+	//speculative execution
+	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	mat2x3 ior = bsdf_data_decodeIoR(bsdf_data,op);
+	params_t params = instr_getParameters(instr, bsdf_data);
+	float bxdf_eval_scalar_part;
+	uint ndf = instr_getNDF(instr);
+	float a = params_getAlpha(params);
+	float a2 = a*a;
+#ifndef ALL_ISOTROPIC_BXDFS
+	float ay = params_getAlphaV(params);
+	float ay2 = ay*ay;
+#endif
+
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
+
+#ifndef NO_TWOSIDED
+	handleTwosided(ts_flag, instr, s, _uf);
+#endif
+
+	const float cosFactor = is_bsdf ? abs(s.NdotL):max(s.NdotL,0.0);
+
+	uvec3 regs = instr_decodeRegisters(instr);
+
+	if (cosFactor<=FLT_MIN && op_isBXDF(op))
+	{
+		writeReg(REG_DST(regs), bxdf_eval_t(0.0));
+		return; //early exit
+	}
+
+	irr_glsl_AnisotropicMicrofacetCache uf;
+	//here actually using stronger check for BSDF because it's probably worth it
+	if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+	{
+		float orientedEta, rcpOrientedEta;
+		irr_glsl_getOrientedEtas(orientedEta, rcpOrientedEta, currInteraction.isotropic.NdotV, ior[0].x);
+		const bool valid = irr_glsl_calcAnisotropicMicrofacetCache(uf, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, orientedEta, rcpOrientedEta);
+		//assert(valid);
```

Comment:
you need to keep track of `valid` and evaluate the BSDF to 0.0 if not valid, see my other comment on `common.glsl` and the eval_pdf function